### ISSUE TITLE
Make OptionsFactory.Create virtual

### ISSUE THREAD

**Issue by hikalkan** [dotnet/extensions#2908]
I had created a PR before (see https://github.com/aspnet/Options/pull/258). It is a simple change for you, but important for us. I [explained it](https://github.com/aspnet/Options/pull/258#issuecomment-406503119) very detailed. Please consider to make this method virtual.

**Comment by anurse** [comment]
Why do you need the method to be virtual? There is an `IOptionsFactory` interface and you can copy what you need from the current implementation out of the code.

Not only is making the method `virtual` breaking change for callers in IL but it introduces a new public contract throughout the type (other methods in the type that call it may get a different implementation than they were prepared to). We don't make methods `virtual` by default as a general principle.

This wouldn't meet the bar for a breaking change. I'm happy to keep the issue open to discuss other options to solve your problem though.

**Comment by hikalkan** [comment]
>Why do you need the method to be virtual?

Re-implementing a complete interface is harder than inheriting a base class especially if you need a minor change. I had to duplicate all the logic (see https://github.com/abpframework/abp/blob/dev/framework/src/Volo.Abp.Core/Volo/Abp/Options/AbpOptionsFactory.cs) in order to make it easily overridable. However, I don't want to duplicate a code inside the framework - the reason is explicit: maintanence. 

>We don't make methods virtual by default as a general principle

I respect, but think differently. Making virtuals allows developers more oppurtunity to customize the framework. At least, make it for current public & protected methods. I would be happier if you could change all privates to protected virtual :)

**Comment by davidfowl** [comment]
If you change the lifetime to scoped doesn’t that affect any implementation that was assuming it was a singleton?

**Comment by hikalkan** [comment]
Yes, I make it scoped ([see](https://github.com/abpframework/abp/blob/81595ee5b37d9ee83e61d95fccbad5167f2a7905/modules/identity/src/Volo.Abp.Identity.Domain/Volo/Abp/Identity/AbpIdentityDomainModule.cs#L71-L75)) and created a custom options factory ([see](https://github.com/abpframework/abp/blob/dev/modules/identity/src/Volo.Abp.Identity.Domain/Volo/Abp/Identity/AbpIdentityOptionsFactory.cs#L13))

This is needed in real life for a multi-tenant system where every tenant may want to have different set of security options.

>doesn’t that affect any implementation that was assuming it was a singleton

I believe a class should not make an assumption on the lifetime of a service it is using. That's not a good design.

It a developer replaces a singleton service by a transient/scoped one, he should care about performance and other aspects. In my implementation, SettingProvider provides a cached service, so no performance problem if you are asking for it.

**Comment by davidfowl** [comment]
That’s not true. The reality is that things are built with lifetimes in mind because objects reference other objects with that assumption. You can’t look at a single dependency in isolation, you need to look at the object graphs that consume those objects. It’s the same reason we have IOptions (singleton) and IOptionsSnapshot (scoped)

**Comment by hikalkan** [comment]
If so, what do you suggest if our `Password.RequiredLength` is changing on runtime? Not only based on multi-tenancy, but we may want to create a page to allow an admin user to change it on the application?

**Comment by davidfowl** [comment]
A feature request for the identity system to work under these requirements to make sure there are no holes.

**Comment by hikalkan** [comment]
>A feature request for the identity system to work under these requirements to make sure there are no holes.

That's good. Whenever it is released, I will drop current workaround solution. Thanks.

**Comment by davidfowl** [comment]
You would need to file that issue on the repository explaining the scenario.

**Comment by hikalkan** [comment]
OK, I will try to find it and join to the discussion. Thanks.
